modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryArrayWrapper.java

```diff
@@ -0,0 +1,204 @@
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Binary object representing array.
+ */
+public class BinaryArrayWrapper implements BinaryObjectEx, Externalizable {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** Context. */
+    @GridDirectTransient
+    private BinaryContext ctx;
+
+    /** Type ID. */
+    private int compTypeId;
+
+    /** Raw data. */
+    private String compClsName;
+
+    /** Value bytes. */
+    private Object[] arr;
+
+    /** Size of the object */
+    private int size;
+
+    /**
+     * {@link Externalizable} support.
+     */
+    public BinaryArrayWrapper() {
+        // No-op.
+    }
+
+    /**
+     * @param ctx Context.
+     * @param compTypeId Component type id.
+     * @param compClsName Component class name.
+     * @param arr Array.
+     * @param size Size of the object.
+     */
+    public BinaryArrayWrapper(BinaryContext ctx, int compTypeId, @Nullable String compClsName, Object[] arr, int size) {
+        this.ctx = ctx;
+        this.compTypeId = compTypeId;
+        this.compClsName = compClsName;
+        this.arr = arr;
+        this.size = size;
+    }
+
+    /** {@inheritDoc} */
+    @Override public BinaryType type() throws BinaryObjectException {
+        return BinaryUtils.typeProxy(ctx, this);
+    }
+
+    /** {@inheritDoc} */
+    @Override public @Nullable BinaryType rawType() throws BinaryObjectException {
+        return BinaryUtils.type(ctx, this);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T> T deserialize() throws BinaryObjectException {
+        return (T)deserialize(null);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T> T deserialize(ClassLoader ldr) throws BinaryObjectException {
+        ClassLoader resolveLdr = ldr == null ? ctx.configuration().getClassLoader() : ldr;
+
+        if (ldr != null)
+            GridBinaryMarshaller.USE_CACHE.set(Boolean.FALSE);
+
+        try {
+            Class compType = BinaryUtils.resolveClass(ctx, compTypeId, compClsName, resolveLdr, false);
+
+            Object[] res = (Object[])Array.newInstance(compType, arr.length);
+
+            for (int i = 0; i < arr.length; i++)
+                    res[i] = arr[i] instanceof BinaryObject ? ((BinaryObject)arr[i]).deserialize(ldr) : arr[i];
+
+            return (T)res;
+        }
+        finally {
+            GridBinaryMarshaller.USE_CACHE.set(Boolean.TRUE);
+        }
+    }
+
+    /**
+     * @return Underlying array.
+     */
+    public Object[] array() {
+        return arr;
+    }
+
+    /**
+     * @return Component type ID.
+     */
+    public int componentTypeId() {
+        return compTypeId;
+    }
+
+    /**
+     * @return Component class name.
+     */
+    public String componentClassName() {
+        return compClsName;
+    }
+
+    /** {@inheritDoc} */
+    @Override public BinaryObject clone() throws CloneNotSupportedException {
+        return new BinaryArrayWrapper(ctx, compTypeId, compClsName, arr.clone(), size);
```

Comment:
IDEA show a warning about using constructor in clone() method, can we use super.clone() here?